### PR TITLE
Add initial GoDaddy tables Closes #1

### LINKED ISSUE #1: Add initial plugin tables
https://github.com/alyx/go-daddy/

- [ ] table_godaddy_domain
- [ ] table_godaddy_dns_recored
- [ ] table_godaddy_subscription
- [ ] table_godaddy_certificate
- [ ] table_godaddy_shopper

### PR DESCRIPTION
**Pull request by rajlearner17**

### CODE REVIEW

**Review by madhushreeray30 — CHANGES_REQUESTED** [review]
Please have a look at the comments. Thanks!!

File: godaddy/table_godaddy_certificate.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+			{
+				Name:        "organization",
+				Description: "Details of the organization that owns the certificate.",
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "subject_alternative_names",
+				Description: "Specifies subject alternative names set for the certificate.",
+				Type:        proto.ColumnType_JSON,
+			},
```
[the remark this answers is not in the document]
Comment (rajlearner17) [reply godaddy/table_godaddy_certificate.go:113]: I added JSON col and example query as an example. The relevance of this column is optional to split it into many columns. Hence it's OK to be there as JSON.

**Review by misraved — COMMENTED** [review]
@ParthaI please take a look at the review comments. Thanks!!

Additional questions:
- Is there no consistency in the not found error codes?  Does the API return a 404 error code?
- Has the pagination been handled correctly? Judging by the API output, I think you could use https://github.com/turbot/steampipe-plugin-launchdarkly/blob/main/launchdarkly/table_launchdarkly_project.go#L134-L154 for reference
- Could you please attach a doc where the max limits of the API calls have been defined?
- Does the API doc mention anything about the rate limit errors?

**Comment by misraved** [comment]
@ParthaI @rajlearner17 I see a total of 10 APIs that can be used to create the tables in - https://developer.godaddy.com/doc

Is there any specific reason why we didn't add other tables?

File: godaddy/table_godaddy_dns_record.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+	name := ""
+	if d.EqualsQualString("name") != "" {
+		name = d.EqualsQualString("name")
+	}
+
+	// To avoid errors, both the input parameters "name" and "DNS type" must be provided simultaneously in the where clause. Passing only one of them will result in an error being thrown by the API.
+	if name == "" || dnsType == "" {
+		name, dnsType = "", ""
+	}
```
Comment (misraved) [inline godaddy/table_godaddy_dns_record.go:155]: Is the logic correct? As per the above code, if either of the parameter is an empty string, then both the values get set to empty.

**Comment by misraved** [comment]
Questions around `environment_type`  and other config arguments -
1. What is the reason behind naming it as `environment_type`? Should it just be `environment` or `base_url`? In my opinion, this should not be a required parameter, rather it should default to production. https://registry.terraform.io/providers/n3integration/godaddy/latest/docs. Please confirm what the API actually defaults to.
2. I feel that the `secret_key` argument should be renamed to `api_secret` as per the sample mentioned below - 
```
Calling the API
Here is a sample call using the API to check if a domain is available. Replace API_KEY and API_SECRETwith your API Key and Secret.

curl -X GET -H "Authorization: sso-key [API_KEY]:[API_SECRET]" "https://api.godaddy.com/v1/domains/available?domain=example.guru"
```

File: godaddy/table_godaddy_dns_record.go (hunk 2 above)
Comment (rajlearner17) [reply godaddy/table_godaddy_dns_record.go:155]: @misraved Those were explored, and we didn't find any relevant data to consider. So we decided to stick to only current table sets.

File: .github/ISSUE_TEMPLATE/feature-request---new-table.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+---
+name: Feature request - New table
+about: Suggest a new table for this project
+title: Add table godaddy <resource>
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply .github/ISSUE_TEMPLATE/feature-request---new-table.md:4]: Updated.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,117 @@
...
+
+Download and install the latest GoDaddy plugin:
+
+```bash
+steampipe plugin install godaddy
+```
+
+Configure your [credentials](https://hub.steampipe.io/plugins/turbot/godaddy#credentials) and [config file](https://hub.steampipe.io/plugins/turbot/godaddy#configuration).
+
+Configure the Organization URL and Personal Access Token in `~/.steampipe/config/godaddy.spc`:
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply README.md:24]: Updated.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,117 @@
...
+  environment_type = "DEV"
+}
+```
+
+Or through environment variables:
+
+```sh
+export GODADDY_API_KEY="hkw64xxxxabchttXW_TmcsFgxJQBvLjE5Lda8402"
+export GODADDY_SECRET_KEY="DjfrsqEBA4vVjsdsdsdieDShdjs"
+export GODADDY_ENVIRONMENT_TYPE="DEV"
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply README.md:42]: Updated

File: docs/index.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+| Radius  | Each connection represents a single GoDaddy account.                                                                                                                             |
+| Resolution  | 1. Credentials explicitly set in a steampipe config file (`~/.steampipe/config/godaddy.spc`)<br />2. Credentials specified in environment variables, e.g., `GODADDY_ENVIRONMENT_TYPE`, `GODADDY_API_KEY` and `GODADDY_SECRET_KEY`.                                                                                                                               |
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply docs/index.md:59]: Re-orderd the environment variable.

File: docs/index.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  # environment_type = "DEV"
+}
+```
+
+Alternatively, you can also use the GoDaddy environment variables to obtain credentials **only if other arguments `api_key`  `secret_key` and `environment_type` are not specified** in the connection:
+
+```sh
+export GODADDY_ENVIRONMENT_TYPE="DEV"
+export GODADDY_API_KEY="hkw64xxxxabchttXW_TmcsFgxJQBvLjE5Lda8402"
+export GODADDY_SECRET_KEY="DjfrsqEBA4vVjsdsdsdieDShdjs"
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply docs/index.md:89]: Re-ordered the list of environment variables

File: docs/tables/godaddy_domain.md [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+  transfer_away_eligibile_at,
+  expires,
+  status
+from
+  godaddy_domain
+where
+  status <> 'ACTIVE';
+```
+
+### List domains that are eligible for transfer away
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply docs/tables/godaddy_domain.md:79]: Updated

File: docs/tables/godaddy_order.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+group by
+  order_id;
+```
+
+### Count the number of orders created per month
+
+```sql
+select
+  order_id,
+  DATE_TRUNC('month', created_at) as month,
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply docs/tables/godaddy_order.md:35]: Updated

File: docs/tables/godaddy_order.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  count(*) as num_orders
+from
+  godaddy_order
+group by
+  order_id,
+  created_at;
+```
+
+### List orders by parent order id
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply docs/tables/godaddy_order.md:44]: Updated

File: docs/tables/godaddy_shopper.md [hunk 11]
```diff
@@ -0,0 +1,23 @@
+# Table: godaddy_shopper
+
+GoDaddy Shopper feature allows customers to create a centralized account to manage all of their GoDaddy products and services in one place.
+
+Note: The used [Go API](https://pkg.go.dev/github.com/alyx/go-daddy/daddy#ShoppersService.Get) limitation, you must provide valid `shopper_id` in the `where` clause of the queries to fetch shopping details.
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply docs/tables/godaddy_shopper.md:5]: Updated

File: docs/tables/godaddy_subscription.md [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+  created_at,
+  cancelable,
+  price_locked
+from
+  godaddy_subscription
+where
+  addons is not null;
+```
+
+### List non cancelable subscriptions
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply docs/tables/godaddy_subscription.md:64]: Updated

File: docs/tables/godaddy_subscription.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+  created_at,
+  expires_at,
+  renew_auto
+from
+  godaddy_subscription
+where
+  created_at between '2022-01-01' and '2022-12-31';
+```
+
+### List subscriptions that has auto renew enabled
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply docs/tables/godaddy_subscription.md:91]: Updated

File: go.mod [hunk 14]
```diff
@@ -0,0 +1,110 @@
+module github.com/turbot/steampipe-plugin-godaddy
+
+go 1.19
+
+require (
+	github.com/turbot/go-daddy v0.0.3
+	github.com/turbot/steampipe-plugin-sdk/v5 v5.3.0
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply go.mod:7]: Updated the SDK version to `v5.4.1`

File: godaddy/connection_config.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+	"github.com/turbot/go-daddy/daddy"
+	"github.com/turbot/steampipe-plugin-sdk/v5/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/v5/plugin/schema"
+)
+
+type godaddyConfig struct {
+	API_KEY         *string `cty:"api_key"`
+	SECRET_KEY      *string `cty:"secret_key"`
+	EVIRONMENT_TYPE *string `cty:"environment_type"`
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply godaddy/connection_config.go:16]: Updated the config parameters

**Comment by ParthaI** [comment]
> What is the reason behind naming it as environment_type? Should it just be environment or base_url? In my opinion, this should not be a required parameter, rather it should default to production. https://registry.terraform.io/providers/n3integration/godaddy/latest/docs. Please confirm what the API actually defaults to.

The [GoDaddy SDK](https://github.com/alyx/go-daddy/blob/master/daddy/godaddy.go#LL186C43-L186C46) which is being used here, it is passing the base URL internally as per the type of GoDaddy environment we have weather we have a Production/Development environment. 

> I feel that the `secret_key` argument should be renamed to `api_secret` as per the sample mentioned below 

Updated the argument `secret_key` to `api_secret`

**Review by misraved — APPROVED** [review]

**Review by misraved — APPROVED** [review]

**Pull request merged**